subxt/src/updates.rs

```diff
@@ -0,0 +1,104 @@
+// Copyright 2019-2022 Parity Technologies (UK) Ltd.
+// This file is part of subxt.
+//
+// subxt is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// subxt is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with subxt.  If not, see <http://www.gnu.org/licenses/>.
+
+//! For performing runtime updates.
+
+use crate::{
+    rpc::{
+        Rpc,
+        RuntimeVersion,
+    },
+    BasicError,
+    Config,
+    Metadata,
+};
+use parking_lot::RwLock;
+use std::sync::Arc;
+
+/// Client wrapper for performing runtime updates.
+pub struct UpdateClient<T: Config> {
+    rpc: Rpc<T>,
+    metadata: Arc<RwLock<Metadata>>,
+    runtime_version: Arc<RwLock<RuntimeVersion>>,
+}
+
+impl<T: Config> UpdateClient<T> {
+    /// Create a new [`UpdateClient`].
+    pub fn new(
+        rpc: Rpc<T>,
+        metadata: Arc<RwLock<Metadata>>,
+        runtime_version: Arc<RwLock<RuntimeVersion>>,
+    ) -> Self {
+        Self {
+            rpc,
+            metadata,
+            runtime_version,
+        }
+    }
+
+    /// Performs runtime updates indefinitely unless encountering an error.
+    ///
+    /// *Note:* This should be called from a dedicated background task.
+    pub async fn perform_runtime_updates(&self) -> Result<(), BasicError> {
+        // Obtain an update subscription to further detect changes in the runtime version of the node.
+        let mut update_subscription = self.rpc.subscribe_runtime_version().await?;
+
+        while let Some(update_runtime_version) = update_subscription.next().await {
+            // The Runtime Version obtained via subscription.
+            let update_runtime_version = update_runtime_version?;
+
+            // Ensure that the provided Runtime Version can be applied to the current
+            // version of the client. There are cases when the subscription to the
+            // Runtime Version of the node would produce spurious update events.
+            // In those cases, set the Runtime Version on the client if and only if
+            // the provided runtime version is bigger than what the client currently
+            // has stored.
```

Comment:
I'd assume that the new runtimeversion would only ever be either identical to the one we currently have (no need to do anything), or different, which implies we should update to match. I'd guess we never need to worry about it being older than what we know about, and if for some reason it is, then we should probably just assume it's true and update to match it anyway?

(basically, maybe we should just do an equality check?)

Reply:
Indeed, that model makes more sense. Have updated the comment as well to describe the spurious event a bit better.
Also, the equality check is something we do for: https://github.com/polkadot-js/api/blob/master/packages/api/src/base/Init.ts#L277. Thanks for pointing this out! :D